feat: add trusted nodes configuration

MegaRedHand opened the pull request:
Linked to: #320
Blocked by: #581
This PR adds the possibility to specify trusted nodes through a list of enodes in the configuration file, along with a flag to deny connection to non-trusted nodes. It also adds the implementation for the first part.

gakonst commented on bin/reth/src/config.rs:
```diff
@@ -78,3 +80,12 @@ impl Default for SenderRecoveryConfig {
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+pub struct DiscoveryConfig {
+    /// Preferred nodes to connect to.
+    pub trusted_nodes: Vec<String>,
```
This should be NodeRecord, and let's impl any missing traits if needed for it

mattsse replied:
yep, I guess we can add `Serialize` Deserialize Support to `NodeRecord` via FromStr + Display

can be a separate PR if you like @MegaRedHand

mattsse replied:
perhaps we even move NodeRecord to primitives

MegaRedHand replied:
I'm going to make these changes in a separate PR then, as per your suggestion.

onbjerg replied:
Imo we should try to not end up copy/pasting the peers config (and network config) to this module as we would end up doing double work. Ideally the network crate would have a `serde` feature to deserialize/serialize the relevant configs. Because of that, I think this particular part should be moved to a separate section of the current config as `peers` will probably clash at some point

gakonst commented on crates/net/discv4/src/config.rs:
```diff
@@ -247,6 +253,21 @@ impl Discv4ConfigBuilder {
     }
 
+    /// Adds multiple preferred nodes
+    pub fn add_preferred_nodes(
```
Let's use consistent naming, if we're calling them trusted in the vars let's call them trusted here

mattsse commented on crates/net/network/src/config.rs:
```diff
@@ -107,6 +107,10 @@ pub struct NetworkConfigBuilder<C> {
+    /// Nodes we always want to connect to.
+    trusted_nodes: Vec<NodeRecord>,
+    /// If we should connect to preferred nodes only.
+    trusted_only: bool,
```
I'd like this to be called `connect_trusted_nodes_only` or similar, should be obvious that this is about connections

MegaRedHand replied:
Should we change it only here, or also in the implementation?

mattsse replied:
everywhere it is used, including function names. there are a lot of peer-related settings and logic, so it's worth being explicit here

mattsse commented on bin/reth/src/node/mod.rs:
```diff
@@ -215,9 +216,12 @@ fn network_config<DB: Database>(
     NetworkConfig::builder(Arc::new(ProviderImpl::new(db)), rng_secret_key())
         .boot_nodes(mainnet_nodes())
+        .trusted_nodes(parse_nodes(config.discovery.trusted_nodes.clone()))
+        .set_trusted_only(false)
```
redundant, since default

mattsse commented on bin/reth/src/config.rs:
```diff
@@ -78,3 +80,12 @@ impl Default for SenderRecoveryConfig {
+
+/// Configuration for each stage in the pipeline.
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+pub struct DiscoveryConfig {
```
hmm, I'm not sure if `Discovery` is semantically fitting here, because these are already "discovered" and will be added to the Peerset directly.

Maybe `Nodes|Peers`Config? I guess `Peers` makes sense, for example there's an rpc endpoint called `admin_addTrustedPeer` which accepts a NodeRecord as string.

mattsse reviewed (commented):
thanks for taking this on.

some early pointers.

MegaRedHand commented:
Just implemented your requested changes. As suggested, I'm gonna add `Serialize` `Deserialize` support and move `NodeRecord` to _primitives_ in another PR.

mattsse commented on bin/reth/src/config.rs:
```diff
@@ -78,3 +80,12 @@ impl Default for SenderRecoveryConfig {
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+pub struct PeersConfig {
+    /// Trusted nodes to connect to.
+    pub trusted_nodes: Vec<String>,
```
should be NodeRecord once we have Serialize/Deserialize
```suggestion
    pub trusted_nodes: Vec<NodeRecord>,
```

MegaRedHand replied:
Changed

mattsse commented on crates/net/discv4/src/config.rs:
```diff
@@ -247,6 +253,18 @@ impl Discv4ConfigBuilder {
         self
     }
 
+    /// Adds multiple trusted nodes
+    pub fn add_trusted_nodes(&mut self, nodes: impl IntoIterator<Item = NodeRecord>) -> &mut Self {
+        self.config.trusted_nodes.extend(nodes);
+        self
+    }
```
nice,

would like to see another helper function for `add_trusted_node(NodeRecord)`

mattsse commented on crates/net/discv4/src/lib.rs:
```diff
@@ -525,6 +525,8 @@ impl Discv4Service {
     pub fn spawn(mut self) -> JoinHandle<()> {
         tokio::task::spawn(async move {
             self.bootstrap();
+            self.add_all_nodes(self.config.trusted_nodes.clone());
+
```
hmm, unsure about how this should be integrated with discovery. since these are different types of nodes, for example perhaps you don't want to announce your trusted nodes via discovery.

So imo should remove here and only use in peerset.

mattsse commented on crates/net/network/src/config.rs:
```diff
@@ -107,6 +107,10 @@ pub struct NetworkConfigBuilder<C> {
     discovery_v4_builder: Discv4ConfigBuilder,
     /// All boot nodes to start network discovery with.
     boot_nodes: Vec<NodeRecord>,
+    /// Nodes we always want to connect to.
+    trusted_nodes: Vec<NodeRecord>,
```
all of these should be HashSet actually.

MegaRedHand replied:
Changed

mattsse commented on crates/net/network/src/config.rs:
```diff
@@ -253,20 +259,34 @@ impl<C> NetworkConfigBuilder<C> {
         self
     }
 
+    /// Sets the trusted nodes.
+    pub fn trusted_nodes(mut self, nodes: impl IntoIterator<Item = NodeRecord>) -> Self {
+        self.trusted_nodes = nodes.into_iter().collect();
+        self
+    }
```
same, +1 `add_` function

MegaRedHand replied:
Renamed to `add_trusted_nodes` and added `add_trusted_node`

mattsse commented on crates/net/network/src/config.rs:
```diff
@@ -297,6 +317,10 @@ impl<C> NetworkConfigBuilder<C> {
             ForkFilter::new(head, genesis_hash, Hardfork::all_forks())
         });
 
+        discovery_v4_builder
+            .add_trusted_nodes(trusted_nodes)
+            .set_connect_trusted_nodes_only(connect_trusted_nodes_only);
```
see above.

mattsse reviewed (changes requested):
great progress,

some nits re config and concern re discovery, otherwise lgtm and we can move on to PeersManager integration.

gakonst reviewed (approved):
Good with me, pending Matt's nits. Agreed let's do the NodeRecord impl/refactor in a follow-up, pls make sure to open an issue so we don't lose track of it!

onbjerg commented on crates/net/discv4/src/config.rs:
```diff
@@ -247,6 +253,18 @@ impl Discv4ConfigBuilder {
+    }
+
+    /// Sets flag for using only trusted nodes
+    pub fn set_connect_trusted_nodes_only(&mut self, trusted_only: bool) -> &mut Self {
+        self.config.connect_trusted_nodes_only = trusted_only;
+        self
+    }
```
Does this sort of clash with #495 @mattsse? In the sense that if we only want to connect to trusted nodes, we should just connect to them via e.g. `NetworkHandle` and skip discv4 entirely, no?

mattsse replied:
yeh, mentioned here https://github.com/paradigmxyz/reth/pull/569#discussion_r1055352612 trusted nodes should not be integrated with Discovery and kept separately.

onbjerg reviewed (changes requested):
I would try to avoid duplicating crate-specific configs in `bin/reth/src/config.rs` and instead implement `Serialize` and `Deserialize` for the configs in the crate hidden behind a `serde` feature, otherwise we will end up having to change config structs in two places every time we make a change to configs in the crates themselves

I also am not sure this really achieves the trusted nodes functionality since it "just" adds the nodes to discovery which does not mean we will connect to them necessarily, and going by #495 discovery might not even be enabled for some people (mostly crate consumers).

I think a better way to connect to some set of trusted nodes would be to connect to them via `NetworkHandle` in `bin/reth/src/node/mod.rs`. Since `NetworkHandle` already gives us the ability to connect to people (without touching discv4 also), I am not sure this feature makes sense to include as a separate thing in the network crate - crate consumers can just use `NetworkHandle` + `parse_nodes`.

mattsse commented:
imo this should be a config option, at least in the `NetworkConfig`

and I agree wrt discovery. trusted nodes should not be involved here.

MegaRedHand commented:
Ok. To summarize:

1. remove trusted nodes (and related flag) from discovery
2. add trusted peers via the `NetworkHandle`
3. add a `serde` feature for the `network` crate, that adds `Serialize` `Deserialize` support for `NetworkConfig`
4. replace `PeersConfig` in `bin/reth` with `NetworkConfig`

Did I understand correctly?
Also, should we address point number 3 in a different PR?

mattsse commented:
yes to 1.

> add trusted peers via the NetworkHandle

no, keep the field in `NetworkConfig`

> Also, should we address point number 3 in a different PR?

yes please!

> replace PeersConfig in bin/reth with NetworkConfig

defer to @onbjerg here

MegaRedHand commented:
Update: 
- removed trusted nodes from discovery config, and moved it to `NetworkConfig`.
- now adds peers inside `NetworkManager::new`
- added changes from main (including serde support for `NodeRecord`)

mattsse commented on crates/net/network/src/manager.rs:
```diff
@@ -176,7 +178,11 @@ where
-        let state = NetworkState::new(client, discovery, peers_manger, genesis_hash);
+        let mut state = NetworkState::new(client, discovery, peers_manger, genesis_hash);
+
+        for NodeRecord { address, tcp_port, udp_port: _, id } in trusted_nodes {
+            state.add_peer_address(id, SocketAddr::from((address, tcp_port)));
+        }
```
this  will create a bunch of useless events which we don't want.

now that I'm seeing this, since trusted nodes are basically only a `PeersManager` thing, perhaps we just move it to the `PeersConfig` or add to the new function of PeersManager.

the former feels cleaner, but is perhaps a bit less convenient in the `NetworkConfig`. However we'll revamp this eventually, so can just move into `PeersConfig` and we'll figure out NetworkConfig ergonomics later.

mattsse commented on crates/net/network/src/peers/manager.rs:
```diff
@@ -127,7 +129,13 @@ impl PeersManager {
             backoff_duration,
+        };
+
+        for NodeRecord { address, tcp_port, udp_port: _, id } in trusted_nodes {
+            manager.add_discovered_node(id, SocketAddr::from((address, tcp_port)));
         }
+
```
again, this will trigger events that we don't want

MegaRedHand replied:
Fixed! Peers are added directly to the hashmap on new, now. Should we add a call to `fill_outbound_slots` after that?

MegaRedHand commented:
@onbjerg The idea is to have `bin/reth`'s _config_ import `NetworkConfig` (and with it `PeersConfig`) from _network_, right? If that's the case, I'll add `serde` to `PeersConfig` (in a different PR) so we can replace the current `PeersConfig` (from `bin/reth`) with the real one.

gakonst commented on bin/reth/src/config.rs:
```diff
@@ -78,3 +83,12 @@ impl Default for SenderRecoveryConfig {
     }
 }
+
+/// Configuration for each stage in the pipeline.
```
Should be specifically about peers Config not for every stage. And yes if we can dedup with other definitions let's try that.

gakonst commented on bin/reth/src/node/mod.rs:
```diff
@@ -213,9 +214,14 @@ fn network_config<DB: Database>(
     db: Arc<DB>,
     chain_id: u64,
     genesis_hash: H256,
+    config: &Config,
```
Would like a network_config(db, genesis) method on Config which returns a Network_config if possible so we don't have many free functions like this.

gakonst replied:
Would also like the NetworkConfig struct to have a spawn_network() method which does what start network does below instead of it being a free fn

MegaRedHand replied:
Added both

gakonst commented on crates/net/network/src/peers/manager.rs:
```diff
@@ -105,15 +105,23 @@ impl PeersManager {
 
+        let mut peers = HashMap::with_capacity(trusted_nodes.len());
+
+        for NodeRecord { address, tcp_port, udp_port: _, id } in trusted_nodes {
```
Nice! Clean.

gakonst reviewed (changes requested):
Nice! Good from me modulo the nits I mentioned. Would also like a small test on the peer manager ensuring the trusted peers are indeed configured correctly to avoid future regressions. Nice work otherwise - thanks for taking the time.

mattsse reviewed (approved):
this is great work.

let's merge and and integrate the trusted node prioritization in a follow up PR
